File: OeipWrapper/OeipCamera.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OeipWrapper
{
    public class OeipCamera
    {
        public int Id { get; private set; } = -1;
        public string DeviceName { get; private set; } = string.Empty;
        public string DeviceId { get; private set; } = string.Empty;

        public bool IsInit
        {
            get
            {
                return this.Id >= 0;
            }
        }

        public List<VideoFormat> VideoFormats { get; private set; } = new List<VideoFormat>();

        //专门搞个字段,是为了避免这个传递给C++的回调会被GC给回收
        protected OnEventDelegate onEventDelegate;
        public event OnEventDelegate OnDeviceEvent;

        protected OnReviceDelegate onReviceDelegate;
        public event OnReviceDelegate OnReviceEvent;

        public OeipCamera()
        {
            onEventDelegate = new OnEventDelegate(OnEventHandle);
            onReviceDelegate = new OnReviceDelegate(OnReviceHandle);
        }

        public void SetDevice(OeipDeviceInfo oeipDevice)
        {
            this.Id = oeipDevice.id;
            this.DeviceId = oeipDevice.deviceId;
            this.DeviceName = oeipDevice.deviceName;
            this.GetCameraFormatList(this.Id);
            OeipHelper.setDeviceDataAction(this.Id, onReviceDelegate);
            OeipHelper.setDeviceEventAction(this.Id, onEventDelegate);
        }

        public int GetFormat()
        {
            if (!IsInit)
                return -1;
            return OeipHelper.getFormat(this.Id);
        }

        public void SetFormat(int index)
        {
            if (!IsInit)
                return;
            OeipHelper.setFormat(Id, index);
        }

        public bool Open()
        {
            if (!IsInit)
                return false;
            return OeipHelper.openDevice(Id);
        }

        public void Close()
        {
            if (!IsInit)
                return;
            OeipHelper.closeDevice(Id);
        }

        public bool IsOpen
        {
            get
            {
                if (IsInit)
                    return OeipHelper.bOpen(Id);
                return false;
            }
        }

        public int FindFormatIndex(int width, int height, int fps = 30)
        {
            if (!IsInit)
                return -1;
            return OeipHelper.findFormatIndex(Id, width, height, fps);
        }

        public void GetCameraFormatList(int index)
        {
            if (!IsInit)
                return;
            int count = OeipHelper.getFormatCount(index);
            if (count > 0)
            {
                var videoFormats = PInvokeHelper.GetPInvokeArray<VideoFormat>(count,
                (IntPtr ptr, int pcount) =>
                {
                    OeipHelper.getFormatList(index, ptr, pcount);
                });
                VideoFormats = videoFormats.ToList();
            }
        }

        public override string ToString()
        {
            if (Id == -1)
                return "none";
            return Id + " " + DeviceName;
        }

        private void OnEventHandle(int type, int code)
        {
            OnDeviceEvent?.Invoke(type, code);
        }

        private void OnReviceHandle(IntPtr data, int width, int height)
        {
            OnReviceEvent?.Invoke(data, width, height);
        }
    }
}

File: OeipUE4/Plugins/OeipPlugins/Source/OeipPlugins/Oeip/OeipCamera.cpp
// Fill out your copyright notice in the Description page of Project Settings.

#include "OeipCamera.h"
#include "OeipManager.h"

using namespace std::placeholders;

OeipCamera::OeipCamera() {
}

OeipCamera::~OeipCamera() {
	formatList.Empty();
	if (bInit()) {
		setDeviceDataHandle(id, nullptr);
		setDeviceEventHandle(id, nullptr);
		Close();
	}
}

bool OeipCamera::bInit() {
	if (id < 0)
		return false;
	return true;
}

void OeipCamera::onDeviceHandle(int type, int code) {
	OnDeviceEvent.Broadcast(type, code);
}

void OeipCamera::onReviceHandle(uint8_t * data, int width, int height) {
	OnDeviceDataEvent.Broadcast(data, width, height);
}

void OeipCamera::SetDevice(FCameraInfo * cameraInfo) {
	id = cameraInfo->index;
	deviceName = cameraInfo->deviceName;
	deviceId = cameraInfo->deviceId;
	if (bInit()) {
		std::function<void(int, int)> deviceHandle = std::bind(&OeipCamera::onDeviceHandle, this, _1, _2);
		setDeviceEventHandle(id, deviceHandle);
		std::function<void(uint8_t*, int, int)> deviceDataHandle = std::bind(&OeipCamera::onReviceHandle, this, _1, _2, _3);
		setDeviceDataHandle(id, deviceDataHandle);
	}
}

int OeipCamera::GetFormat() {
	if (!bInit())
		return false;
	return getFormat(id);
}

void OeipCamera::SetFormat(int index) {
	if (!bInit())
		return;
	setFormat(id, index);
}

bool OeipCamera::Open() {
	if (!bInit())
		return false;
	return openDevice(id);
}

void OeipCamera::Close() {
	if (!bInit())
		return;
	closeDevice(id);
}

bool OeipCamera::IsOpen() {
	if (!bInit())
		return false;
	return bOpen(id);
}

int OeipCamera::FindFormatIndex(int width, int height, int fps) {
	if (!bInit())
		return -1;
	return findFormatIndex(id, width, height, fps);
}

TArray<VideoFormat> OeipCamera::GetFormatList() {
	if (bInit() && formatList.Num() == 0) {
		formatList = OeipManager::Get().GetCameraFormatList(id);
	}
	return formatList;
}
